### PR TITLE
fix: fix the rollout stuck issue when pod/replicas changed together for canary rollout. 

### PR DESCRIPTION
**Pull request by andyliuliming**
when the rollout is in the "Pause" status,
and replica/poc spec changed together, then the rollout will stuck there forever.
this is because the newRS will be nil,
and there's one short circut return in reconcileNewReplicaSet().

issue reported in:
https://github.com/argoproj/argo-rollouts/issues/3256

the change will follow the "BlueGreen" style to handle this.
pick the replica set which the canary svc targeted to. and reconcile it.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by andyliuliming** [comment]
will add ut later.

File: rollout/service.go [hunk 1]
```diff
@@ -284,6 +285,27 @@ func (c *rolloutContext) reconcileStableAndCanaryService() error {
 	return nil
 }
 
+func (c *rolloutContext) getCanaryReplicaSet() (*appsv1.ReplicaSet, error) {
+	if c.rollout.Spec.Strategy.Canary == nil {
+		return nil, nil
+	}
+	svc, err := c.servicesLister.Services(c.rollout.Namespace).Get(c.rollout.Spec.Strategy.Canary.CanaryService)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
```
Comment (carolwu-1206) [inline rollout/service.go:294]: If there is no canaryService, the rollout will be stuck too?

Comment (andyliuliming) [reply rollout/service.go:294]: yes, but that's a negative case.

but this is really a goot catch! i should check whether the canary service is an empty string first.

**Comment by zachaller** [comment]
Linking to the BlueGreen issue for reference: https://github.com/argoproj/argo-rollouts/pull/3060

**Comment by zachaller** [comment]
@andyliuliming I could still be doing something wrong so still digging but this PR does not seem to fix the issue for me.

**Comment by zachaller** [comment]
I also think the issue is around this code:

```
func (c *rolloutContext) rolloutCanary() error {
	var err error
	if replicasetutil.PodTemplateOrStepsChanged(c.rollout, c.newRS) {
		c.newRS, err = c.getAllReplicaSetsAndSyncRevision(false)
		if err != nil {
			return err
		}
		return c.syncRolloutStatusCanary()
	}

	c.newRS, err = c.getAllReplicaSetsAndSyncRevision(true)
	if err != nil {
		return err
	}
```

**Comment by zachaller** [comment]
This also happens in Rollouts 1.5

**Comment by andyliuliming** [comment]
@zachaller 

> @andyliuliming I could still be doing something wrong so still digging but this PR does not seem to fix the issue for me.

this only works when there's canaryservice defined in rollout.
the fix works as expected in our rollout.

**Comment by zachaller** [comment]
You are right I don't know how I missed that in my test rollout. Still going to spend some time on it though because I feel there is a more underlying issue, I want to investigate that a bit more but this does seem to fix the issue as long as you have canary service defined.

**Comment by zachaller** [comment]
I think this would also fix the issue but also work for rollouts without a canary service: https://github.com/argoproj/argo-rollouts/pull/3272/files

**Comment by andyliuliming** [comment]
> I think this would also fix the issue but also work for rollouts without a canary service: https://github.com/argoproj/argo-rollouts/pull/3272/files

tested. it works!

**Pull request closed without merging**